odo push does not fail if executed devfile command failed to start

/kind bug

<!--

Welcome! - We kindly ask you to:

  1. Fill out the issue template below 
  2. Use the Google group if you have a question rather than a bug or feature request.

The group is at: https://groups.google.com/forum/#!forum/odo-users

Thanks for understanding, and for contributing to the project!

-->

## What versions of software are you using?

**Operating System:** Linux

**Output of `odo version`:** 2.0.1

## How did you run odo exactly?
```
user@host:~$ git clone https://github.com/dgolovin/nodejs-ex ./nodejs-devfile-issue
Cloning into './nodejs-devfile-issue'...
remote: Enumerating objects: 637, done.
remote: Total 637 (delta 0), reused 0 (delta 0), pack-reused 637
Receiving objects: 100% (637/637), 253.59 KiB | 2.59 MiB/s, done.
Resolving deltas: 100% (246/246), done.

user@host:~$ cd nodejs-devfile-issue/

user@host:~/nodejs-devfile-issue$ odo create nodejs --app app1 --project p1
Validation
 ✓  Checking devfile existence [35224ns]
 ✓  Checking devfile compatibility [55663ns]
 ✓  Creating a devfile component from registry: DefaultDevfileRegistry [55176ns]
 ✓  Validating devfile component [94611ns]

Please use `odo push` command to create the component with source deployed
```
## Actual behavior

```
user@host:~/nodejs-devfile-issue$  odo push --debug

Validation
 ✓  Validating the devfile [59256ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [6s]

Applying URL changes
 ✓  URL http-3000: http://http-3000-nodejs-p1.apps-crc.testing/ created

Syncing to component nodejs
 ✓  Checking files for pushing [677080ns]
 ✓  Syncing files to the component [223ms]

Executing devfile commands for component nodejs
 ✓  Executing install command "npm install" [5s]
 ✓  Executing debug command "npm run debug" [1s]

Pushing devfile component nodejs
 ✓  Changes successfully pushed to component

user@host:~/nodejs-devfile-issue$ odo log
time="2020-11-05T09:06:56Z" level=info msg="create process:devrun" 
time="2020-11-05T09:06:56Z" level=info msg="create process:debugrun" 
time="2020-11-05T09:07:02Z" level=debug msg="no auth required" 
time="2020-11-05T09:07:02Z" level=debug msg="succeed to find process:debugrun" 
time="2020-11-05T09:07:02Z" level=info msg="try to start program" program=debugrun 
time="2020-11-05T09:07:02Z" level=info msg="success to start program" program=debugrun 
ODO_COMMAND_DEBUG is npm run debug
Changing directory to /project
Executing command cd /project && npm run debug
npm ERR! missing script: debug

npm ERR! A complete log of this run can be found in:
npm ERR!     /opt/app-root/src/.npm/_logs/2020-11-05T09_07_03_073Z-debug.log
time="2020-11-05T09:07:03Z" level=debug msg="wait program exit" program=debugrun 
time="2020-11-05T09:07:03Z" level=error msg="program stopped with error:exit status 1" program=debugrun 
time="2020-11-05T09:07:03Z" level=info msg="Don't start the stopped program because its retry times 0 is greater than start retries 0" program=debugrun
```

Odo push does not fails even if executed devfile command does.

## Expected behavior

If executing devfile command fails, odo push fails.

## Any logs, error output, etc?

examle how this could look like for normal output:

```
▶ odo push

Validation
 ✓  Validating the devfile [126100ns]

Creating Kubernetes resources for component nodejs
 ✓  Waiting for component to start [5s]
 ✓  Waiting for component to start [171ms]

Applying URL changes
 ✓  URL http-3000: http://http-3000-nodejs-tkral-test1.apps.testocp4.psiodo.net/ created

Syncing to component nodejs
 ✓  Checking files for pushing [930000ns]
 ✓  Syncing files to the component [3s]

Executing devfile commands for component nodejs
 ✓  Waiting for component to start [182ms]
 ✓  Executing install command "npm install" [3s]
 ✗  Executing run command "npm start" [2s]
 ✗  Command "npm start" exited with error status within 1s.

time="2021-03-05T08:02:54Z" level=info msg="create process:devrun"
time="2021-03-05T08:02:54Z" level=info msg="create process:debugrun"
time="2021-03-05T08:03:03Z" level=debug msg="no auth required"
time="2021-03-05T08:03:03Z" level=debug msg="succeed to find process:devrun"
time="2021-03-05T08:03:03Z" level=info msg="try to start program" program=devrun
time="2021-03-05T08:03:03Z" level=info msg="success to start program" program=devrun
ODO_COMMAND_RUN is npm start
Changing directory to /project
Executing command cd /project && npm start

> nodejs-starter@1.0.0 start /project
> node-fail server.js

sh: node-fail: command not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! nodejs-starter@1.0.0 start: `node-fail server.js`
npm ERR! spawn ENOENT
npm ERR!
npm ERR! Failed at the nodejs-starter@1.0.0 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /opt/app-root/src/.npm/_logs/2021-03-05T08_03_03_542Z-debug.log
time="2021-03-05T08:03:04Z" level=debug msg="wait program exit" program=devrun
time="2021-03-05T08:03:04Z" level=error msg="program stopped with error:exit status 1" program=devrun
time="2021-03-05T08:03:04Z" level=info msg="Don't start the stopped program because its retry times 0 is greater than start retries 0" program=devrun

Pushing devfile component nodejs
 ✗  Error during component startup.
```

--- a/pkg/devfile/adapters/common/interface.go
+++ b/pkg/devfile/adapters/common/interface.go
@@ -2,6 +2,8 @@ package common
 
 import (
 	"io"
+
+	devfilev1 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
 )
 
 // ComponentAdapter defines the functions that platform-specific adapters must implement
@@ -11,8 +13,9 @@ type ComponentAdapter interface {
 	DoesComponentExist(cmpName string) (bool, error)
 	Delete(labels map[string]string, show bool) error
 	Test(testCmd string, show bool) error
+	CheckSupervisordCtlStatus(command devfilev1.Command) error
 	StartContainerStatusWatch()
 	StartSupervisordCtlStatusWatch()
-	Log(follow, debug bool) (io.ReadCloser, error)
+	Log(follow bool, command devfilev1.Command) (io.ReadCloser, error)
 	Exec(command []string) error
 }

--- a/pkg/devfile/adapters/docker/adapter.go
+++ b/pkg/devfile/adapters/docker/adapter.go
@@ -38,6 +38,10 @@ func (d Adapter) Push(parameters common.PushParameters) error {
 	return nil
 }
 
+func (a Adapter) CheckSupervisordCtlStatus(command devfilev1.Command) error {
+	return nil
+}
+
 // DoesComponentExist returns true if a component with the specified name exists
 func (d Adapter) DoesComponentExist(cmpName string) (bool, error) {
 	return d.componentAdapter.DoesComponentExist(cmpName)
@@ -54,8 +58,8 @@ func (d Adapter) Test(testCmd string, show bool) error {
 }
 
 // Log shows logs from component
-func (d Adapter) Log(follow, debug bool) (io.ReadCloser, error) {
-	return d.componentAdapter.Log(follow, debug)
+func (d Adapter) Log(follow bool, command devfilev1.Command) (io.ReadCloser, error) {
+	return d.componentAdapter.Log(follow, command)
 }
 
 // Exec executes a command in the component

--- a/pkg/devfile/adapters/docker/component/adapter.go
+++ b/pkg/devfile/adapters/docker/component/adapter.go
@@ -3,7 +3,6 @@ package component
 import (
 	"fmt"
 	"io"
-	"reflect"
 	"strings"
 
 	"github.com/docker/docker/api/types"
@@ -182,6 +181,10 @@ func (a Adapter) Push(parameters common.PushParameters) (err error) {
 	return nil
 }
 
+func (a Adapter) CheckSupervisordCtlStatus(command devfilev1.Command) error {
+	return nil
+}
+
 // Test runs the devfile test command
 func (a Adapter) Test(testCmd string, show bool) (err error) {
 	componentExists, err := utils.ComponentExists(a.Client, a.Devfile.Data, a.ComponentName)
@@ -346,7 +349,7 @@ func (a Adapter) Delete(labels map[string]string, show bool) error {
 }
 
 // Log returns log from component
-func (a Adapter) Log(follow, debug bool) (io.ReadCloser, error) {
+func (a Adapter) Log(follow bool, command devfilev1.Command) (io.ReadCloser, error) {
 
 	exists, err := utils.ComponentExists(a.Client, a.Devfile.Data, a.ComponentName)
 
@@ -363,23 +366,6 @@ func (a Adapter) Log(follow, debug bool) (io.ReadCloser, error) {
 		return nil, errors.Wrapf(err, "error while retrieving container for odo component %s", a.ComponentName)
 	}
 
-	var command devfilev1.Command
-	if debug {
-		command, err = common.GetDebugCommand(a.Devfile.Data, "")
-		if err != nil {
-			return nil, err
-		}
-		if reflect.DeepEqual(devfilev1.Command{}, command) {
-			return nil, errors.Errorf("no debug command found in devfile, please run \"odo log\" for run command logs")
-		}
-
-	} else {
-		command, err = common.GetRunCommand(a.Devfile.Data, "")
-		if err != nil {
-			return nil, err
-		}
-
-	}
 	containerID := utils.GetContainerIDForAlias(containers, command.Exec.Component)
 
 	return a.Client.GetContainerLogs(containerID, follow)

--- a/pkg/devfile/adapters/kubernetes/adapter.go
+++ b/pkg/devfile/adapters/kubernetes/adapter.go
@@ -42,6 +42,16 @@ func (k Adapter) Push(parameters common.PushParameters) error {
 	return nil
 }
 
+// CheckSupervisordCtlStatus calls the component adapter's CheckSupervisordCtlStatus
+func (k Adapter) CheckSupervisordCtlStatus(command devfilev1.Command) error {
+	err := k.componentAdapter.CheckSupervisordCtlStatus(command)
+	if err != nil {
+		return errors.Wrap(err, "Failed to check the status")
+	}
+
+	return nil
+}
+
 // DoesComponentExist returns true if a component with the specified name exists
 func (k Adapter) DoesComponentExist(cmpName string) (bool, error) {
 	return k.componentAdapter.DoesComponentExist(cmpName)
@@ -64,8 +74,8 @@ func (k Adapter) Test(testCmd string, show bool) error {
 }
 
 // Log shows log from component
-func (k Adapter) Log(follow, debug bool) (io.ReadCloser, error) {
-	return k.componentAdapter.Log(follow, debug)
+func (k Adapter) Log(follow bool, command devfilev1.Command) (io.ReadCloser, error) {
+	return k.componentAdapter.Log(follow, command)
 }
 
 // Exec executes a command in the component

--- a/pkg/devfile/adapters/kubernetes/component/adapter.go
+++ b/pkg/devfile/adapters/kubernetes/component/adapter.go
@@ -3,8 +3,10 @@ package component
 import (
 	"fmt"
 	"io"
+	"os"
 	"reflect"
 	"strings"
+	"time"
 
 	"github.com/devfile/library/pkg/devfile/generator"
 	componentlabels "github.com/openshift/odo/pkg/component/labels"
@@ -34,6 +36,8 @@ import (
 	"github.com/openshift/odo/pkg/sync"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 )
+
+const supervisorDStatusWaitTimeInterval = 1
 
 // New instantiates a component adapter
 func New(adapterContext common.AdapterContext, client kclient.Client) Adapter {
@@ -248,12 +252,66 @@ func (a Adapter) Push(parameters common.PushParameters) (err error) {
 		if err != nil {
 			return err
 		}
+
+		runCommand := pushDevfileCommands[devfilev1.RunCommandGroupKind]
+		if parameters.Debug {
+			runCommand = pushDevfileCommands[devfilev1.DebugCommandGroupKind]
+		}
+
+		// wait for a second
+		wait := time.After(supervisorDStatusWaitTimeInterval * time.Second)
+		<-wait
+
+		err := a.CheckSupervisordCtlStatus(runCommand)
+		if err != nil {
+			return err
+		}
 	} else {
 		// no file was modified/added/deleted/renamed, thus return without syncing files
 		log.Success("No file changes detected, skipping build. Use the '-f' flag to force the build.")
 	}
 
 	return nil
+}
+
+// CheckSupervisordCtlStatus checks the supervisord status according to the given command
+// if the command is not in a running state, we fetch the last 20 lines of the component's log and display it
+func (a Adapter) CheckSupervisordCtlStatus(command devfilev1.Command) error {
+	statusInContainer := getSupervisordStatusInContainer(a.pod.Name, command.Exec.Component, a)
+
+	supervisordProgramName := "devrun"
+
+	// if the command is a debug one, we check against `debugrun`
+	if command.Exec.Group.Kind == devfilev1.DebugCommandGroupKind {
+		supervisordProgramName = "debugrun"
+	}
+
+	for _, status := range statusInContainer {
+		if strings.ToLower(status.program) == strings.ToLower(supervisordProgramName) {
+			if strings.ToLower(status.status) == strings.ToLower("running") {
+				return nil
+			} else {
+				numberOfLines := 20
+				log.Errorf("devfile command \"%s\" exited with error status within %d sec", command.Id, supervisorDStatusWaitTimeInterval)
+				log.Infof("Last %d lines of the component's log:", numberOfLines)
+
+				rd, err := a.Log(false, command)
+				if err != nil {
+					return err
+				}
+
+				err = util.DisplayLog(false, rd, os.Stderr, a.ComponentName, numberOfLines)
+				if err != nil {
+					return err
+				}
+
+				log.Info("To get the full log output, please run 'odo log'")
+
+				return fmt.Errorf("error during component startup")
+			}
+		}
+	}
+	return fmt.Errorf("the supervisord program %s not found", supervisordProgramName)
 }
 
 // Test runs the devfile test command
@@ -530,7 +588,7 @@ func (a Adapter) Delete(labels map[string]string, show bool) error {
 }
 
 // Log returns log from component
-func (a Adapter) Log(follow, debug bool) (io.ReadCloser, error) {
+func (a Adapter) Log(follow bool, command devfilev1.Command) (io.ReadCloser, error) {
 
 	pod, err := a.Client.GetPodUsingComponentName(a.ComponentName)
 	if err != nil {
@@ -539,23 +597,6 @@ func (a Adapter) Log(follow, debug bool) (io.ReadCloser, error) {
 
 	if pod.Status.Phase != corev1.PodRunning {
 		return nil, errors.Errorf("unable to show logs, component is not in running state. current status=%v", pod.Status.Phase)
-	}
-
-	var command devfilev1.Command
-	if debug {
-		command, err = common.GetDebugCommand(a.Devfile.Data, "")
-		if err != nil {
-			return nil, err
-		}
-		if reflect.DeepEqual(devfilev1.Command{}, command) {
-			return nil, errors.Errorf("no debug command found in devfile, please run \"odo log\" for run command logs")
-		}
-
-	} else {
-		command, err = common.GetRunCommand(a.Devfile.Data, "")
-		if err != nil {
-			return nil, err
-		}
 	}
 
 	containerName := command.Exec.Component

--- a/pkg/devfile/adapters/kubernetes/component/adapter.go
+++ b/pkg/devfile/adapters/kubernetes/component/adapter.go
@@ -287,8 +287,8 @@ func (a Adapter) CheckSupervisordCtlStatus(command devfilev1.Command) error {
 	}
 
 	for _, status := range statusInContainer {
-		if strings.ToLower(status.program) == strings.ToLower(supervisordProgramName) {
-			if strings.ToLower(status.status) == strings.ToLower("running") {
+		if strings.EqualFold(status.program, supervisordProgramName) {
+			if strings.EqualFold(status.status, "running") {
 				return nil
 			} else {
 				numberOfLines := 20

--- a/pkg/devfile/adapters/kubernetes/component/adapter.go
+++ b/pkg/devfile/adapters/kubernetes/component/adapter.go
@@ -292,7 +292,7 @@ func (a Adapter) CheckSupervisordCtlStatus(command devfilev1.Command) error {
 				return nil
 			} else {
 				numberOfLines := 20
-				log.Errorf("devfile command \"%s\" exited with error status within %d sec", command.Id, supervisorDStatusWaitTimeInterval)
+				log.Warningf("devfile command \"%s\" exited with error status within %d sec", command.Id, supervisorDStatusWaitTimeInterval)
 				log.Infof("Last %d lines of the component's log:", numberOfLines)
 
 				rd, err := a.Log(false, command)
@@ -307,7 +307,7 @@ func (a Adapter) CheckSupervisordCtlStatus(command devfilev1.Command) error {
 
 				log.Info("To get the full log output, please run 'odo log'")
 
-				return fmt.Errorf("error during component startup")
+				return nil
 			}
 		}
 	}

--- a/pkg/occlient/deploymentConfigs.go
+++ b/pkg/occlient/deploymentConfigs.go
@@ -3,6 +3,7 @@ package occlient
 import (
 	"encoding/json"
 	"fmt"
+	"os"
 	"sort"
 	"time"
 
@@ -230,7 +231,7 @@ func (c *Client) DisplayDeploymentConfigLog(deploymentConfigName string, followL
 		return errors.New("unable to retrieve DeploymentConfig from OpenShift, does your component exist?")
 	}
 
-	return util.DisplayLog(followLog, rd, deploymentConfigName)
+	return util.DisplayLog(followLog, rd, os.Stdout, deploymentConfigName, -1)
 }
 
 // StartDeployment instantiates a given deployment

--- a/pkg/odo/cli/component/devfile.go
+++ b/pkg/odo/cli/component/devfile.go
@@ -3,8 +3,10 @@ package component
 import (
 	"os"
 	"path/filepath"
+	"reflect"
 	"strings"
 
+	devfilev1 "github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
 	"github.com/devfile/library/pkg/devfile"
 	"github.com/openshift/odo/pkg/devfile/validate"
 	"github.com/openshift/odo/pkg/envinfo"
@@ -183,8 +185,25 @@ func (lo LogOptions) DevfileComponentLog() error {
 		return err
 	}
 
+	var command devfilev1.Command
+	if lo.debug {
+		command, err = common.GetDebugCommand(devObj.Data, "")
+		if err != nil {
+			return err
+		}
+		if reflect.DeepEqual(devfilev1.Command{}, command) {
+			return errors.Errorf("no debug command found in devfile, please run \"odo log\" for run command logs")
+		}
+
+	} else {
+		command, err = common.GetRunCommand(devObj.Data, "")
+		if err != nil {
+			return err
+		}
+	}
+
 	// Start or update the component
-	rd, err := devfileHandler.Log(lo.logFollow, lo.debug)
+	rd, err := devfileHandler.Log(lo.logFollow, command)
 	if err != nil {
 		log.Errorf(
 			"Failed to log component with name %s.\nError: %v",
@@ -194,7 +213,7 @@ func (lo LogOptions) DevfileComponentLog() error {
 		os.Exit(1)
 	}
 
-	return util.DisplayLog(lo.logFollow, rd, componentName)
+	return util.DisplayLog(lo.logFollow, rd, os.Stdout, componentName, -1)
 }
 
 // DevfileComponentDelete deletes the devfile component

--- a/pkg/util/util.go
+++ b/pkg/util/util.go
@@ -1197,7 +1197,8 @@ func addFileToIgnoreFile(gitIgnoreFile, filename string, fs filesystem.Filesyste
 }
 
 // DisplayLog displays logs to user stdout with some color formatting
-func DisplayLog(followLog bool, rd io.ReadCloser, compName string) (err error) {
+// numberOfLastLines limits the number of lines from the output when we are not following it
+func DisplayLog(followLog bool, rd io.ReadCloser, writer io.Writer, compName string, numberOfLastLines int) (err error) {
 
 	defer rd.Close()
 
@@ -1217,12 +1218,11 @@ func DisplayLog(followLog bool, rd io.ReadCloser, compName string) (err error) {
 			os.Exit(1)
 		}()
 
-		if _, err = io.Copy(os.Stdout, rd); err != nil {
+		if _, err = io.Copy(writer, rd); err != nil {
 			return errors.Wrapf(err, "error followLoging logs for %s", compName)
 		}
 
-	} else {
-
+	} else if numberOfLastLines == -1 {
 		// Copy to buffer (we aren't going to be followLoging the logs..)
 		buf := new(bytes.Buffer)
 		_, err = io.Copy(buf, rd)
@@ -1231,10 +1231,36 @@ func DisplayLog(followLog bool, rd io.ReadCloser, compName string) (err error) {
 		}
 
 		// Copy to stdout
-		if _, err = io.Copy(os.Stdout, buf); err != nil {
+		if _, err = io.Copy(writer, buf); err != nil {
 			return errors.Wrapf(err, "error copying logs to stdout")
 		}
+	} else {
+		reader := bufio.NewReader(rd)
+		var lines []string
+		for {
+			line, err := reader.ReadString('\n')
+			if err != nil {
+				if err != io.EOF {
+					return err
+				} else {
+					break
+				}
+			}
 
+			lines = append(lines, line)
+		}
+
+		index := len(lines) - numberOfLastLines
+		if index < 0 {
+			index = 0
+		}
+
+		for i := index; i < len(lines)-1; i++ {
+			_, err := fmt.Fprintf(writer, lines[i])
+			if err != nil {
+				return err
+			}
+		}
 	}
 	return
 

--- a/tests/integration/devfile/cmd_devfile_debug_test.go
+++ b/tests/integration/devfile/cmd_devfile_debug_test.go
@@ -183,13 +183,13 @@ var _ = Describe("odo devfile debug command tests", func() {
 			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-with-debugrun.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
 			helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "npm run debug", "npm run debugs")
 
-			output := helper.CmdShouldFail("odo", "push", "--debug", "--context", commonVar.Context)
+			_, output := helper.CmdShouldPassIncludeErrStream("odo", "push", "--debug", "--context", commonVar.Context)
 			helper.MatchAllInOutput(output, []string{
 				"exited with error status within 1 sec",
 				"Did you mean this?",
 			})
 
-			output = helper.CmdShouldFail("odo", "push", "--debug", "--context", commonVar.Context, "--debug-command", "debug")
+			_, output = helper.CmdShouldPassIncludeErrStream("odo", "push", "--debug", "--context", commonVar.Context, "--debug-command", "debug", "-f")
 			helper.MatchAllInOutput(output, []string{
 				"exited with error status within 1 sec",
 				"Did you mean this?",

--- a/tests/integration/devfile/cmd_devfile_debug_test.go
+++ b/tests/integration/devfile/cmd_devfile_debug_test.go
@@ -175,4 +175,25 @@ var _ = Describe("odo devfile debug command tests", func() {
 
 		})
 	})
+
+	Context("when the debug command throws an error during push", func() {
+		It("should wait and error out with some log", func() {
+			helper.CmdShouldPass("odo", "create", "nodejs", "--project", commonVar.Project, "--context", commonVar.Context)
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), commonVar.Context)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-with-debugrun.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
+			helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "npm run debug", "npm run debugs")
+
+			output := helper.CmdShouldFail("odo", "push", "--debug", "--context", commonVar.Context)
+			helper.MatchAllInOutput(output, []string{
+				"exited with error status within 1 sec",
+				"Did you mean this?",
+			})
+
+			output = helper.CmdShouldFail("odo", "push", "--debug", "--context", commonVar.Context, "--debug-command", "debug")
+			helper.MatchAllInOutput(output, []string{
+				"exited with error status within 1 sec",
+				"Did you mean this?",
+			})
+		})
+	})
 })

--- a/tests/integration/devfile/cmd_devfile_push_test.go
+++ b/tests/integration/devfile/cmd_devfile_push_test.go
@@ -1090,13 +1090,13 @@ var _ = Describe("odo devfile push command tests", func() {
 			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
 			helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "npm start", "npm starts")
 
-			output := helper.CmdShouldFail("odo", "push", "--context", commonVar.Context)
+			_, output := helper.CmdShouldPassIncludeErrStream("odo", "push", "--context", commonVar.Context)
 			helper.MatchAllInOutput(output, []string{
 				"exited with error status within 1 sec",
 				"Did you mean one of these?",
 			})
 
-			output = helper.CmdShouldFail("odo", "push", "--context", commonVar.Context, "-f", "--run-command", "run")
+			_, output = helper.CmdShouldPassIncludeErrStream("odo", "push", "--context", commonVar.Context, "-f", "--run-command", "run")
 			helper.MatchAllInOutput(output, []string{
 				"exited with error status within 1 sec",
 				"Did you mean one of these?",

--- a/tests/integration/devfile/cmd_devfile_push_test.go
+++ b/tests/integration/devfile/cmd_devfile_push_test.go
@@ -1083,6 +1083,27 @@ var _ = Describe("odo devfile push command tests", func() {
 		})
 	})
 
+	Context("when the run command throws an error", func() {
+		It("should wait and error out with some log", func() {
+			helper.CmdShouldPass("odo", "create", "nodejs", "--project", commonVar.Project, "--context", commonVar.Context)
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), commonVar.Context)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
+			helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "npm start", "npm starts")
+
+			output := helper.CmdShouldFail("odo", "push", "--context", commonVar.Context)
+			helper.MatchAllInOutput(output, []string{
+				"exited with error status within 1 sec",
+				"Did you mean one of these?",
+			})
+
+			output = helper.CmdShouldFail("odo", "push", "--context", commonVar.Context, "-f", "--run-command", "run")
+			helper.MatchAllInOutput(output, []string{
+				"exited with error status within 1 sec",
+				"Did you mean one of these?",
+			})
+		})
+	})
+
 	Context("Testing Push for OpenShift specific scenarios", func() {
 		JustBeforeEach(func() {
 			if os.Getenv("KUBERNETES") == "true" {

